Extend reflection configuration for SmallRye FT

zakkak opened the pull request:
Doesn't seem to fix any functionality issues but prevents
`MissingRegistrationErrors` from being thrown when
`ThrowMissingRegistrationErrors` is enabled.

Related to https://github.com/quarkusio/quarkus/issues/41995

zakkak committed:
Extend reflection configuration for SmallRye FT

Related to https://github.com/quarkusio/quarkus/issues/41995

Ladicek reviewed (approved):
LGTM for the fault tolerance change. Someone else should approve the Vert.x changes, although they seem like a no-brainer.

gsmet reviewed (dismissed):
CI looks unhappy:

```
2024-07-30T20:52:39.8530794Z Caused by: org.graalvm.nativeimage.MissingReflectionRegistrationError: The program tried to reflectively invoke method public java.lang.String io.quarkus.it.rest.client.main.FaultToleranceClient$$CDIWrapper.fallback() without it being registered for runtime reflection. Add public java.lang.String io.quarkus.it.rest.client.main.FaultToleranceClient$$CDIWrapper.fallback() to the reflection metadata to solve this problem. See https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection for help.
2024-07-30T20:52:39.8536648Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.forQueriedOnlyExecutable(MissingReflectionRegistrationUtils.java:72)
2024-07-30T20:52:39.8538942Z 	at java.base@21.0.4/java.lang.reflect.Method.acquireMethodAccessor(Method.java:77)
2024-07-30T20:52:39.8540282Z 	at java.base@21.0.4/java.lang.reflect.Method.invoke(Method.java:577)
2024-07-30T20:52:39.8542250Z 	at io.smallrye.faulttolerance.core.invocation.NormalMethodInvoker.proceed(NormalMethodInvoker.java:46)
2024-07-30T20:52:39.8544470Z 	at io.smallrye.faulttolerance.FaultToleranceInterceptor.lambda$prepareFallbackFunction$13(FaultToleranceInterceptor.java:575)
2024-07-30T20:52:39.8546131Z 	... 27 more
```

Ladicek commented:
OK, that's weird. Since this PR only _adds_ more reflection config to fault tolerance, that suggests there's more reflection configuration for the same class and they are not combined properly :shrug:

zakkak commented:
> CI looks unhappy:
> 
> ```
> 2024-07-30T20:52:39.8530794Z Caused by: org.graalvm.nativeimage.MissingReflectionRegistrationError: The program tried to reflectively invoke method public java.lang.String io.quarkus.it.rest.client.main.FaultToleranceClient$$CDIWrapper.fallback() without it being registered for runtime reflection. Add public java.lang.String io.quarkus.it.rest.client.main.FaultToleranceClient$$CDIWrapper.fallback() to the reflection metadata to solve this problem. See https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection for help.
> 2024-07-30T20:52:39.8536648Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.forQueriedOnlyExecutable(MissingReflectionRegistrationUtils.java:72)
> 2024-07-30T20:52:39.8538942Z 	at java.base@21.0.4/java.lang.reflect.Method.acquireMethodAccessor(Method.java:77)
> 2024-07-30T20:52:39.8540282Z 	at java.base@21.0.4/java.lang.reflect.Method.invoke(Method.java:577)
> 2024-07-30T20:52:39.8542250Z 	at io.smallrye.faulttolerance.core.invocation.NormalMethodInvoker.proceed(NormalMethodInvoker.java:46)
> 2024-07-30T20:52:39.8544470Z 	at io.smallrye.faulttolerance.FaultToleranceInterceptor.lambda$prepareFallbackFunction$13(FaultToleranceInterceptor.java:575)
> 2024-07-30T20:52:39.8546131Z 	... 27 more
> ```

`io.quarkus.it.rest.client.main.FaultToleranceClient$$CDIWrapper` is a generated class that looks like this:

```java
@ApplicationScoped
@RestClient
@Typed({FaultToleranceClient.class})
@Path("/unprocessable")
public class FaultToleranceClient$$CDIWrapper extends RestClientReactiveCDIWrapperBase implements FaultToleranceClient {
   public FaultToleranceClient$$CDIWrapper() {
      ClassLoader var1 = Thread.currentThread().getContextClassLoader();
      Class var2 = Class.forName("io.quarkus.it.rest.client.main.FaultToleranceClient", false, var1);
      super(var2, "", "w-fault-tolerance", false);
   }

   @Consumes({"text/plain"})
   @Produces({"text/plain"})
   public String hello() {
      return ((FaultToleranceClient)this.getDelegate()).hello();
   }

   public String fallback() {
      return super.fallback();
   }

   @Fallback(
      fallbackMethod = "fallback"
   )
   public String helloWithFallback() {
      return super.helloWithFallback();
   }
}
```

Since `helloWithFallback` is annotated with `@Fallback`, the `fallback` method (i.e. the value of the annotation) should be registered for reflection (see https://github.com/quarkusio/quarkus/blob/c9edbe64520c0c906a5b481017c44b37e267c3b2/extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java#L124-L159) but this is not happening.

I suspect/guess that the reason it's not happening is because the generated class is not indexed when processing the annotated methods.

@Ladicek @gsmet any clues how I can overcome this and get the method registered?

zakkak commented:
> OK, that's weird. Since this PR only _adds_ more reflection config to fault tolerance, that suggests there's more reflection configuration for the same class and they are not combined properly 🤷

The added reflection seems to change the smallrye behavior though, since `io.smallrye.faulttolerance.internal.SecurityActions.findDeclaredMethodNames` used to fail with an exception before, but now it doesn't.

I wonder if this is a common pattern in places where the recorder is used, i.e. the recorder ensures things will be properly set up and work, while the library code silently fails when trying to configure things but without breaking things thanks to the recorder (similarly to https://github.com/quarkusio/quarkus/pull/42140#discussion_r1692972087).

mkouba reviewed (approved):
The Vert.x changes LGTM.

Ladicek commented on extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java:
```diff
@@ -287,6 +250,49 @@ void processFaultToleranceAnnotations(SmallRyeFaultToleranceRecorder recorder,
+            }
+            String fallbackMethod = fallbackMethodValue.asString();
+
+            System.out.println("Looking for fallback method: " + fallbackMethod);
```
Should be removed I guess?

zakkak replied:
Yes, I will switch this to draft once the CI completes since there seem to be other issues as well.

Ladicek replied:
Ah, OK, understood.

Ladicek commented on extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java:
```diff
@@ -287,6 +250,49 @@ void processFaultToleranceAnnotations(SmallRyeFaultToleranceRecorder recorder,
+                    continue;
+                }
+
+                System.out.println("Looking into class: " + clazz);
```
Same here.

zakkak committed:
Extend ReflectiveClassBuildItem with queryPublicMethods option

Configures whether declared methods should be registered for reflection,
for query purposes only, i.e. {@link Class#getMethods()}. Setting this
enables getting all declared methods for the class but does not allow
invoking them reflectively.

Follow up to https://github.com/quarkusio/quarkus/pull/42035

zakkak commented:
This is now ready for review.

Some comments:

1. It looks like there are multiple places were Quarkus gets away without the "proper" reflection configuration. As stated in https://github.com/quarkusio/quarkus/pull/42229#issuecomment-2260175197 I wonder if the approach I am taking is the right one, i.e. registering the missing stuff, or whether we should substitute the code performing these reflective accesses as Quarkus seems to work OK without them. I believe we will need some domain expert to decide on this on a per case bases. @Ladicek can you please have a look on this PR and provide some feedback regarding the above as far as FT is concerned? Thanks

2. It is based on https://github.com/quarkusio/quarkus/pull/42265 (so please review only the commits specific to this PR)

Ladicek commented on extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java:
```diff
@@ -260,6 +220,67 @@ public void transform(TransformationContext ctx) {
+                }
+                name = classInfo.superName();
+            }
+            clazz.interfaceNames().forEach(it -> classesToScan.computeIfAbsent(it, k -> new HashSet<>()).add(fallbackMethod));
```
This means that only methods from interfaces implemented directly by the bean class are registered, but not methods from interfaces implemented by superclasses of the bean class.

zakkak replied:
Correct, looking at the smallrye code my understanding is that it only checks those, but maybe I missed something.

See https://github.com/smallrye/smallrye-fault-tolerance/blob/c0b05c97b37f5824150af2f9e4ec09ae42b93ea9/implementation/fault-tolerance/src/main/java/io/smallrye/faulttolerance/internal/SecurityActions.java#L131-L207

Now that I looked at it again the code seems to indeed check only the interfaces of the bean class but the comments mention that:

```
        // if we find a matching method on the bean class or one of its superclasses or superinterfaces,
        // then we have to check that the method is either identical to or an override of a method that:
        // - is declared on a class which is a supertype of the declaring class, or
        // - is declared on an interface which implemented by the declaring class
```

which implies that the code should scan the superinterfaces of the bean class as well but it doesn't?

Ladicek replied:
Good point, that is indeed a bug in SmallRye Fault Tolerance.

zakkak replied:
OK, then I will adjust my PR (in anticipation of this getting fixed in SmallRye FT as well). Can you please take care of the SmallRye FT issue and post the relevant issue/PR link here for reference? I could do it but I guess you already have a fork of the repo :)

Ladicek replied:
Sure, I'm fixing it at the moment.

Ladicek replied:
Fix is in https://github.com/smallrye/smallrye-fault-tolerance/pull/1045

zakkak commented on extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java:
```diff
@@ -260,6 +220,79 @@ public void transform(TransformationContext ctx) {
+            // Scan both the hierarchy of the declaring class and its interfaces like in
+            // io.smallrye.faulttolerance.internal.SecurityActions.findDeclaredMethodNames
+            DotName name = clazz.name();
+            boolean isBean = beanArchiveIndexBuildItem.getGeneratedClassNames().contains(name);
```
@Ladicek is this is the right way to do it?

Could we have non-generated beans as well or generated classes that are not actually beans? 

If so, how can I most effectively check if the DotName corresponds to a bean?

Ladicek replied:
I'm not sure why you'd need to do this. If you find a `@Fallback` annotation in the bean archive index, the declaring class is most likely a bean class (and if not, it is most likely a superclass of one). I wouldn't bother trying to figure out whether the class is a bean class or not.

zakkak replied:
I am trying to keep the reflection registrations to the minimum and in theory we only need to register the superinterfaces of beans, so if something is not a bean (and also not a superclass of a bean) I am trying to skip registering its superinterfaces.

Is it safe to assume that having a non-bean class with a `@Fallback` annotated method is  **impossible** or at least **highly unlikely**?

I might be overthinking it, but I am not familiar with the code so please bear with me :)

Ladicek replied:
Is is technically possible to have a class with `@Fallback` that is not a bean class, but it's most likely a bug. I would classify that as _highly unlikely_, as you eloquently put it :-)

zakkak replied:
Check removed PTAL

gsmet commented:
@Ladicek if you are still happy with the latest changes, I let you merge this one, thanks!

Sanne commented:
Would it be possible to only register things for reflection as last resort?

I feel we should always try, in order:
1. patch the library to not need reflection
2. patch the extension to try resolve the reflection needs at build time, e.g. "run init during build" (and avoid runtime reflection)
3. register for reflection conditionally: only register the cases which are being used by the actual application instance (and couldn't be avoided in other ways)

Ladicek commented:
I'm not sure this is done, actually. I'd prefer to wait for @zakkak to return from PTO before doing anything with this PR.

zakkak commented:
After discussing this with @Ladicek we concluded that:

1. https://github.com/quarkusio/quarkus/issues/41995 is a "nice to have" at the moment, but might become a "must have" in the future. The issues I am trying to resolve with this PR are due to the `--exact-reachability-metadata` flag, so they are not "actual" issues for the time being.
2. It is possible to do the work currently requiring reflection at build time, but hasn't been done yet due to not being trivial.
3. At some point it might even be possible to completely avoid using reflection, by moving to CDI method invokers and getting rid of reflection registration altogether, but @Ladicek is afraid we can't do that yet due to non-Quarkus usage
4. It looks like putting the effort to do the work at build time would result in "better" native code, but there is no hard evidence it is required at the moment. The increased footprint due to the reflection usage in this case is not that significant (if we add `--exact-reachability-metadata` to the mix then things get more interesting though).

With the above in mind I propose I move the VertX changes in a different PR and keep this PR as draft with just the smallrye-FT additions for future reference, hoping that eventually it might not be needed.

Adding the metadata for all methods to be able to be queried doesn't affect correctness in this case, but just avoids the exception thrown when using `--exact-reachability-metadata`.

Ladicek commented:
> With the above in mind I propose I move the VertX changes in a different PR and keep this PR as draft with just the smallrye-FT additions for future reference

Agreed.

zakkak committed:
Extend reflection configuration for VertX

Doesn't seem to fix any functionality issues but prevents
`MissingRegistrationErrors` from being thrown when
`ThrowMissingRegistrationErrors` is enabled.

Related to quarkusio#41995

zakkak commented:
Superseded by https://github.com/quarkusio/quarkus/pull/43833

Thanks @Ladicek !
